src/ZoneServer/World/Actors/Characters/Components/PersonalStorageComponent.cs

```diff
@@ -0,0 +1,512 @@
+			addedPosition = -1;
+
+			if (CheckStorageFull())
+				return StorageResult.StorageFull;
+
+			// Add to new position
+			var availablePosition = this.FindAvailablePosition();
+			if (this.Add(item, availablePosition, out var addedAmount) == StorageResult.Success)
+			{
+				addedPosition = availablePosition;
+				return StorageResult.Success;
+			}
+
+			return StorageResult.InvalidOperation;
+		}
+
+		/// <summary>
+		/// Adds a new item to the storage in a specific position.
+		/// Does not update client.
+		/// </summary>
+		/// <param name="item">Item to add</param>
+		/// <param name="position">Position of item in storage to add to.</param>
+		/// <param name="addedAmount">Amount that was added. Defaults to 0</param>
+		/// <returns></returns>
+		public StorageResult Add(Item item, int position, out int addedAmount)
+		{
+			addedAmount = 0;
+
+			if ((position < 0) || (position >= _maxStorage))
+				return StorageResult.InvalidOperation;
+
+			var existingItem = this.TryGetItemPosition(position);
+			if (existingItem == null)
+			{
+				// Adds item
+				lock (_syncLock)
+				{
+					_storageItems[position] = item;
+					_occupiedPositions += 1;
+				}
+				addedAmount = item.Amount;
+				return StorageResult.Success;
+			}
+			else
+			{
+				// Stacks item
+				if (existingItem.IsStackable && item.IsStackable)
+				{
+					if (existingItem.Data.ClassName == item.Data.ClassName)
+					{
+						lock (_syncLock)
+							addedAmount = AddItemStack(existingItem, item.Amount);
+						return StorageResult.Success;
+					}
+				}
+			}
+
+			return StorageResult.InvalidOperation;
+		}
+
+		/// <summary>
+		/// Removes an item from storage.
+		/// Does not update client.
+		/// </summary>
+		/// <param name="item">Item to remove</param>
+		/// <param name="amount">Amount of the item</param>
+		/// <param name="removedPosition">Position where item was removed. Defaults to -1</param>
+		/// <param name="removedAmount">Amount that was removed. Defaults to 0</param>
+		/// <returns></returns>
+		public StorageResult Remove(Item item, int amount, out int removedPosition, out int removedAmount)
+		{
+			removedAmount = 0;
+			removedPosition = -1;
+
+			if (amount <= 0)
+				return StorageResult.InvalidOperation;
+
+			var result = this.TryGetItem(item.ObjectId, out item, out removedPosition);
+			if (result != StorageResult.Success)
+				return result;
+
+			// If the item is stackable, handle the reduction of the stack
+			if (item.IsStackable)
+			{
+				lock (_syncLock)
+					removedAmount = RemoveItemStack(item, amount, removedPosition);
+			}
+			else
+			{
+				lock (_syncLock)
+				{
+					_storageItems[removedPosition] = null;
+					_occupiedPositions -= 1;
+					removedAmount = 1;
+				}
+			}
+			return StorageResult.Success;
+		}
+
+		/// <summary>
+		/// Returns true if storage is full.
+		/// </summary>
+		/// <returns></returns>
+		public bool CheckStorageFull()
+		{
+			lock (_syncLock)
+				return _occupiedPositions == _maxStorage;
```

Comment:
Yeah I added that as some sort of optimization so you don't have to iterate through the array looking for the item but I was also wondering if that was even necessary since it is basically O(1) anyways.